// environment/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isWebWorker = exports.isSharedWorker = exports.isServiceWorker = exports.isNode = exports.isJsDom = exports.isElectron = exports.isDeno = exports.isDedicatedWorker = exports.isBun = exports.isBrowser = void 0;
/* eslint-disable n/prefer-global/process */
/* globals WorkerGlobalScope, DedicatedWorkerGlobalScope, SharedWorkerGlobalScope, ServiceWorkerGlobalScope */

const isBrowser = exports.isBrowser = globalThis.window?.document !== undefined;
const isNode = exports.isNode = globalThis.process?.versions?.node !== undefined;
const isBun = exports.isBun = globalThis.process?.versions?.bun !== undefined;
const isDeno = exports.isDeno = globalThis.Deno?.version?.deno !== undefined;
const isElectron = exports.isElectron = globalThis.process?.versions?.electron !== undefined;
const isJsDom = exports.isJsDom = globalThis.navigator?.userAgent?.includes('jsdom') === true;
const isWebWorker = exports.isWebWorker = typeof WorkerGlobalScope !== 'undefined' && globalThis instanceof WorkerGlobalScope;
const isDedicatedWorker = exports.isDedicatedWorker = typeof DedicatedWorkerGlobalScope !== 'undefined' && globalThis instanceof DedicatedWorkerGlobalScope;
const isSharedWorker = exports.isSharedWorker = typeof SharedWorkerGlobalScope !== 'undefined' && globalThis instanceof SharedWorkerGlobalScope;
const isServiceWorker = exports.isServiceWorker = typeof ServiceWorkerGlobalScope !== 'undefined' && globalThis instanceof ServiceWorkerGlobalScope;
