// get-stream/source/string.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getStreamAsString = getStreamAsString;
var _contents = require("./contents.js");
var _utils = require("./utils.js");
async function getStreamAsString(stream, options) {
  return (0, _contents.getStreamContents)(stream, stringMethods, options);
}
const initString = () => ({
  contents: '',
  textDecoder: new TextDecoder()
});
const useTextDecoder = (chunk, {
  textDecoder
}) => textDecoder.decode(chunk, {
  stream: true
});
const addStringChunk = (convertedChunk, {
  contents
}) => contents + convertedChunk;
const truncateStringChunk = (convertedChunk, chunkSize) => convertedChunk.slice(0, chunkSize);
const getFinalStringChunk = ({
  textDecoder
}) => {
  const finalChunk = textDecoder.decode();
  return finalChunk === '' ? undefined : finalChunk;
};
const stringMethods = {
  init: initString,
  convertChunk: {
    string: _utils.identity,
    buffer: useTextDecoder,
    arrayBuffer: useTextDecoder,
    dataView: useTextDecoder,
    typedArray: useTextDecoder,
    others: _utils.throwObjectStream
  },
  getSize: _utils.getLengthProp,
  truncateChunk: truncateStringChunk,
  addChunk: addStringChunk,
  getFinalChunk: getFinalStringChunk,
  finalize: _utils.getContentsProp
};
